tool/tsh/access_request.go

```diff
@@ -0,0 +1,308 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"time"
+	"unicode"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/asciitable"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/services"
+
+	"github.com/gravitational/trace"
+)
+
+func onRequestList(cf *CLIConf) error {
+	tc, err := makeClient(cf, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if cf.Username == "" {
+		cf.Username = tc.Username
+	}
+
+	var reqs []types.AccessRequest
+
+	err = tc.WithRootClusterClient(cf.Context, func(clt auth.ClientI) error {
+		reqs, err = clt.GetAccessRequests(cf.Context, types.AccessRequestFilter{})
+		return trace.Wrap(err)
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if cf.ReviewableRequests {
+		filtered := reqs[:0]
+	Reviewable:
+		for _, req := range reqs {
+			if req.GetUser() == cf.Username {
+				continue Reviewable
+			}
+			for _, rev := range req.GetReviews() {
+				if rev.Author == cf.Username {
+					continue Reviewable
+				}
+			}
+			filtered = append(filtered, req)
+		}
+		reqs = filtered
+	}
+	if cf.SuggestedRequests {
+		filtered := reqs[:0]
+	Suggested:
+		for _, req := range reqs {
+			if req.GetUser() == cf.Username {
+				continue Suggested
+			}
+			for _, rev := range req.GetReviews() {
+				if rev.Author == cf.Username {
+					continue Suggested
+				}
+			}
+			for _, reviewer := range req.GetSuggestedReviewers() {
+				if reviewer == cf.Username {
+					filtered = append(filtered, req)
+					continue Suggested
+				}
+			}
+		}
+		reqs = filtered
+	}
+	if cf.MyRequests {
+		filtered := reqs[:0]
```

Comment:
Fine as-is.  While it isn't particularly *useful* to combine the currently supported filters (i.e. a combination of `--my-requests` and `--reviewable` always returns an empty list since users can't review their own requests), there is nothing inherently invalid about making that query.  Just so happens that those filters have no overlap.  Since future iterations might support filters which are useful to combine (e.g. `--created-after`), I figure its best to have filters be combinable from the beginning.

Reply:
To me it looks as if not all filters are inherently combinable which is evident with the combination of  e.g. `--reviewable` and `--my-requests`. While `--reviewable` and `--suggested-reviews` do not exclude each other, it might still be confusing to the user as to which takes precedence. But I could be exaggerating.